designs-for-feedback/trigger-workflow-in-pipeline.md

```diff
@@ -0,0 +1,38 @@
+
+## Proposed solution
+1. We will add a parameter to the triggering of pipelines (`POST /project/:project_slug/pipeline`) to specify a particular workflow by name.
+2. We will also add a new way to specify in config to not run a particular workflow unless it is explicitly triggered.
+
+## Design options for the new triggering parameter.
+In v1.1 we combined the custom parameters passed into a build with our built-in parameters, passing them all when triggering a job as a single name/value collection. In v2 we plan to separate out your custom parameters from built-in flags and settings.
+
+When you want to trigger an individual workflow in a configuration file we can either:
+
+1. Add a new, optional triggering parameter called `workflow` or `workflow_name` that will be passed in the body of your POST when triggering a pipeline. The benefits of this approach is that it's explicit about the intent, easy to read in code, and straight-forward to debug. The potential downside is that if and when we add the ability to trigger an individual job or add other specialized triggering instructions it will start to create a crowded set of new built-in parameters.
+2.  Add a new, optional triggering parameter called something like `scope` or `trigger-scope` that takes a complex object that would look something like `{workflow-name: foo}`. The benefits of this is that we could include things like the branch or tag or SHA within that object, unifying all the various trigger-time settings into a single, extensible object. The downside is added syntactic complexity (and the commensurate cognitive/documentation complexity) and more open-ended debugging.
+
+Working assumption: we will go with option 1 above.
+
+## Design options for configuration of workflows
+If a pipeline configuration contains multiple workflows the behavior today is to run them both immediately upon triggering. With the ability to trigger a particular workflow via the API, users will likely want to be able to configure workflows intended to be run only when explicitly triggered and not run when a normal webhook from GitHub or other VCS system is received upon a push to the repo. Thus, we need to introduce some mechanism to declare under what circumstances certain workflows in a configuration should be run.
+
+Working assumption is that we will add a new valid value like `explicit` or `manual` for the existing `triggers` stanza to only trigger the workflow when it's explicitly invoked. That might look like one of the following:
+
+```
+workflows:
+  foo:
+    triggers:
+      - explicit
+```
+
+The above would prevent the workflow `foo` from being triggered unless it was invoked explicitly when the pipeline is triggered.
+
+## Questions
+
+1. Is there any known use-case where we need to allow triggering multiple workflows vs. one? Assumption: no, if people want to trigger multiple workflows they can send multiple triggers.
```

Comment:
I think monorepos would fall into this bucket, we have some repos ourselves with multiple independent workflows where it might make sense to trigger more than one.

As you say, it will always be possible to trigger multiple pipelines with one workflow each so this isn't _that_ important but it might be nice for the monorepo case to group them into one pipeline.

Reply:
In theory we could have one for now and use a comma-delimited list as the value later. AFAIK you can't have commas in workflow names. That's a bit messy, but probably OK. Alternatively, we could say if it's a simple string it's one and allow `[foo,bar]` later.

Reply:
Is there any downside for designing this to be a real list from the get-go?

Reply:
The input can be either JSON (in which case lists are easy) or URL form-encoded which is... not so pleasant.

Reply:
If we allow JSON later, we can go with a single value for now. That's simpler to build, document, and debug.

Reply:
I meant that right now we support JSON and form-encoded parameters for request bodies. JSON is definitely the easier one for us to work with on the backend.